Deep Stubs Incompatible With Mocking Enum

AndreasTu opened the pull request:
Mockito can't mock abstract enums in Java 15 or later because they are now marked as sealed.
So Mockito reports that now with a better error message.

If a deep stub returns an abstract enum, it uses in the error case now the first enum literal of the real enum.

Fixes #2984

TimvdLippe commented on build.gradle:
```diff
@@ -140,3 +140,50 @@ spotless {
+ * SourceSet for Java 17 tests
+ */
+sourceSets {
+    test17 {
```
Instead of a new sourceset on the main project, let's define a separate subproject for this. Instead of setting to 17 (which is non-LTS), let's use 21 for this. E.g. create `subprojects/java21`. We can use it for #3167 as well

TimvdLippe replied:
As posted in #3167 let's create a separate to introduce a Java 21 subproject so that we can coordinate it (and not do double-work in these two PRs)

TimvdLippe commented on src/main/java/org/mockito/internal/stubbing/defaultanswers/ReturnsDeepStubs.java:
```diff
@@ -76,8 +77,16 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
+        try {
+            return deepStub(invocation, returnTypeGenericMetadata);
+        } catch (MockitoCantMockEnumException ex) {
+            // If the enum is not mockable, e.g. due to sealed, return the first literal as value
```
Rather than swallowing the exception here, I would prefer to always throw the exception and let the user to decide what to return. E.g. you can use `DEEP_STUBS` as default answer and then override this one stub to specify the enum you expect to return. I don't think defaulting to the first value is expected behavior from users.

AndreasTu commented:
@TimvdLippe I have fixed your remarks. And I have also moved the spotless code into `gradle/spottless.gradle` to activate the spotless checks also in the new `java21` subproject.

TimvdLippe commented on subprojects/java21/java21.gradle:
```diff
@@ -1,5 +1,7 @@
 apply from: "$rootDir/gradle/dependencies.gradle"
 apply from: "$rootDir/gradle/java-test.gradle"
+apply from: "$rootDir/gradle/spotless.gradle"
```
Can we do the spotless fixes in a separate PR? Good find btw!

In that PR, let's instead add it to `java-library` and `java-test` so that all the other Java subprojects also have it.

AndreasTu replied:
I will create the other PR.

TimvdLippe commented on subprojects/java21/src/test/java/org/mockito/internal/stubbing/answers/DeepStubReturnsEnumJava21Test.java:
```diff
@@ -0,0 +1,146 @@
+                            mock(TestEnum.class);
+                        })
+                .isInstanceOf(MockitoException.class)
+                .hasMessage(
+                        "\nMockito cannot mock this class: class org.mockito.internal.stubbing.answers.DeepStubReturnsEnumJava21Test$TestEnum.\n"
+                                + "Sealed abstract enums can't be mocked. Since Java 15 abstract enums are declared sealed, which prevents mocking.\n"
+                                + "You can still return an existing enum literal from a stubbed method call.")
+                .hasCauseInstanceOf(MockitoException.class);
```
```suggestion
                .hasMessageContaining("Sealed abstract enums can't be mocked.")
```

This is sufficient to test it throws the correct. It allows us to more easily tweak the error message without having to constantly update the tests as well.

TimvdLippe replied:
Same with the other exception assertions

TimvdLippe commented on subprojects/java21/src/test/java/org/mockito/internal/stubbing/answers/DeepStubReturnsEnumJava21Test.java:
```diff
@@ -0,0 +1,146 @@
+                            mock.getTestEnum();
+                        })
+                .isInstanceOf(MockitoException.class)
+                .hasMessageContaining("Sealed abstract enums can't be mocked.");
```
Ah here you did it already. Yeah basically this, but with the other two test methods

TimvdLippe reviewed (changes requested):
Couple of nits in the test and let's do the spotless work in a different PR and apply to all subprojects. Most likely that will cause a bit more reformatting.

TimvdLippe reviewed (approved):
Thanks!
